### PR TITLE
V2.1.0

### PR DESCRIPTION
**Pull request by aewhite**
@sawyercade review

Changes:

* Support for expunging shards from a blacklisted node
* First class support for node exclusion at the simulation level (faster/better simulations) 
* Made all tempest settings dynamic 
* The _tempest endpoint now works correctly on non-primary masters
* Added a "status" field to _tempest status output
* Removed cluster score, risk, and ratio from status output (pretty unless to the average user)
* Added a _tempest/rebalance endpoint for forcing a rebalance without tweaking settings
* Removed structural hashing and time based backoffs for balancing (too fragile with dynamic settings)

### CODE REVIEW

File: src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/TempestShardsAllocator.kt [hunk 1]
```diff
@@ -24,55 +24,88 @@
...
         return HeuristicBalancer(
-                settings,
+                effectiveSettings,
                 allocation,
                 shardSizeCalculator,
-                balancerState,
-                Random()).rebalance();
+                balancerConfiguration,
+                Random()).rebalance().apply {
+            if (this == true) {
```
Comment (sawyercade) [inline src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/TempestShardsAllocator.kt]: Euh. I get what's happening here, but the syntax is pretty foreign. Is the whitespacing correct/aligned properly?

File: src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/balancer/BalancerConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            expungeBlacklistedNodes = settings.getAsBoolean(TempestConstants.EXPUNGE_BLACKLISTED_NODES, other.expungeBlacklistedNodes),
+            clusterExcludeFilter = buildBlacklistFilter(settings, other.clusterExcludeFilter))
+
+
+}
+
+private fun buildBlacklistFilter(settings: Settings, defaultFilter: (RoutingNode) -> Boolean): (RoutingNode) -> Boolean =
+        settings.getByPrefix(FilterAllocationDecider.CLUSTER_ROUTING_EXCLUDE_GROUP).getAsMap()
+                .let {
+                    if (it.isEmpty()) defaultFilter
```
Comment (sawyercade) [inline src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/balancer/BalancerConfiguration.kt]: No `{ }` ? I'm chill with it, just curious why we require them for one-liner if-elses in java but not kotlin. Is Kotlin just smarter about parsing if-elses to not need them?

File: src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/balancer/ModelCluster.kt [hunk 3]
```diff
@@ -130,30 +155,27 @@ class ModelCluster private constructor(
...
     /**
      * Calculate the balance score for the entire cluster
+     *
+     * NOTE: for blacklisted nodes while an expunge is happening we treat their capacity as 0 since we are trying to
+     *       move everything off of them
      */
-    fun calculateBalanceScore(): Double = modelNodes.fold(0.0, { score, node -> score + node.calculateNodeScore(expectedUnitCapacity) })
+    fun calculateBalanceScore(): Double = modelNodes.fold(
+            0.0,
+            { score, node -> score + node.calculateNodeScore(expectedUnitCapacity) })
```
Comment (sawyercade) [inline src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/balancer/ModelCluster.kt]: `andrew.interpret()` ? Where is `score` coming from? The `0.0` above?

File: src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/TempestShardsAllocator.kt (hunk 1 above)
Comment (aewhite) [inline src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/TempestShardsAllocator.kt]: I'm still experimenting with the indenting here. If this looks wonky I might try a deeper indent. IntelliJ still doesn't have good defaults for Kotlin in my opinion.

File: src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/balancer/ModelCluster.kt (hunk 3 above)
Comment (aewhite) [inline src/main/java/com/simplymeasured/elasticsearch/plugins/tempest/balancer/ModelCluster.kt]: The `fold` function (and this is really a left-fold) says start with a value (0.0 in this case) and evaluate the function for each element in the list and use the output of the evaluation as the input to the next call in the chain.

**Comment by aewhite** [comment]
Your assumption about expungeBlacklisted nodes is correct.

**Pull request merged**